dali/kernels/shape.h

```diff
@@ -0,0 +1,609 @@
+  static_assert(other_ndim != DynamicDimensions, "This function can produce only static shapes");
+  assert(0 <= other_ndim && other_ndim <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<other_ndim> result;
+  result.shapes.resize(other_ndim * get_size());
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < other_ndim; d++) {
+      result.shapes[sample * other_ndim + d] = shapes[sample * get_sample_dim() + d];
+    }
+  }
+  return result;
+}
+
+template <typename Derived, int sample_ndim>
+template <int other_ndim>
+TensorListShape<other_ndim> TensorListShapeBase<Derived, sample_ndim>::last() {
+  static_assert(other_ndim <= sample_ndim || sample_ndim == DynamicDimensions,
+                "Number of dimensions in subshape must be between 0 and sample_dim()");
+  static_assert(other_ndim != DynamicDimensions, "This function can produce only static shapes");
+  assert(0 <= other_ndim && other_ndim <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<other_ndim> result;
+  result.shapes.resize(other_ndim * get_size());
+  int start_offset = get_sample_dim() - other_ndim;
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < other_ndim; d++) {
+      result.shapes[sample * other_ndim + d] = shapes[sample * get_sample_dim() + start_offset + d];
+    }
+  }
+  return result;
+}
+
+template <typename Derived, int sample_ndim>
+TensorListShape<DynamicDimensions> TensorListShapeBase<Derived, sample_ndim>::first(int count) {
+  assert(0 <= count && count <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<DynamicDimensions> result;
+  result.shapes.resize(count * get_size());
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < count; d++) {
+      result.shapes[sample * count + d] = shapes[sample * get_sample_dim() + d];
+    }
+  }
+  return result;
+}
+
+template <typename Derived, int sample_ndim>
+TensorListShape<DynamicDimensions> TensorListShapeBase<Derived, sample_ndim>::last(int count) {
+  assert(0 <= count && count <= get_sample_dim() &&
+         "Number of dimensions in subshape must be between 0 and sample_dim()");
+  TensorListShape<DynamicDimensions> result;
+  result.shapes.resize(count * get_size());
+  int start_offset = get_sample_dim() - count;
+  for (int sample = 0; sample < get_size(); sample++) {
+    for (int d = 0; d < count; d++) {
+      result.shapes[sample * count + d] = shapes[sample * get_sample_dim() + start_offset + d];
+    }
+  }
+  return result;
+}
+
+template <int left_ndim, int right_ndim>
+bool operator==(const TensorListShape<left_ndim> &left, const TensorListShape<right_ndim> &right) {
+  if (left.sample_dim() != right.sample_dim()) {
+    return false;
+  }
+  if (left.size() != right.size()) {
```

Comment:
I think it's redundant - `left.shapes == right.shapes` will do it for us, I think.

Reply:
Done.